File: reference_code/world_model/model/tracker_model/tracker_model.cpp
#include "tracker_model.h"

namespace gpal::pnc::planning {

TrackerModel::TrackerModel() : OptimalControlProblem("TrackerModel", 5, 2, 1, 28, 3) {
  default_state_.emplace("s");
  default_state_.emplace("x");
  default_state_.emplace("y");
  default_state_.emplace("theta");
  default_state_.emplace("steer");

  default_ctrl_.emplace("dsteer");
  default_ctrl_.emplace("l_slack");

  default_param_.emplace("x_ref", 0.0);
  default_param_.emplace("y_ref", 0.0);
  default_param_.emplace("theta_ref", 0.0);
  default_param_.emplace("kappa_ref", 0.0);
  default_param_.emplace("v_ref", 0.0);
  default_param_.emplace("l_terminal", 0.0);
  default_param_.emplace("wheelbase", 3.0);
  default_param_.emplace("front_overhang", 1.0);
  default_param_.emplace("rear_overhang", 1.0);
  default_param_.emplace("length", 5.0);
  default_param_.emplace("width", 2.0);
  default_param_.emplace("l_ref_weight", 1.0);
  default_param_.emplace("theta_ref_weight", 1.0);
  default_param_.emplace("steer_weight", 1.0);
  default_param_.emplace("dsteer_weight", 1.0);
  default_param_.emplace("l_slack_weight", 1.0);
  default_param_.emplace("terminal_l_weight", 1.0);
  default_param_.emplace("terminal_theta_weight", 1.0);
  default_param_.emplace("LHardLowerBound", -20.0);
  default_param_.emplace("LHardUpperBound", 20.0);
  default_param_.emplace("SteerLowerBound", -0.436);
  default_param_.emplace("SteerUpperBound", 0.436);
  default_param_.emplace("DSteerLowerBound", -0.436);
  default_param_.emplace("DSteerUpperBound", 0.436);
  default_param_.emplace("l_scale", 1.0);
  default_param_.emplace("theta_scale", 1.0);
  default_param_.emplace("steer_scale", 1.0);
  default_param_.emplace("dsteer_scale", 1.0);

  default_global_.emplace("kappa");
  default_global_.emplace("l");
  default_global_.emplace("ds");
  default_global_.setUpdater(std::make_shared<OcpVariable::Updater>([](const OcpStage& stage, OcpVariable* ptr_global) {
    // Determine stage variables
    const auto& x = stage.x(1);
    const auto& y = stage.x(2);
    const auto& theta = stage.x(3);
    const auto& steer = stage.x(4);
    const auto& x_ref = stage.params(0);
    const auto& y_ref = stage.params(1);
    const auto& theta_ref = stage.params(2);
    const auto& kappa_ref = stage.params(3);
    const auto& wheelbase = stage.params(6);

    (*ptr_global)(0) = tan(steer)/wheelbase;
    (*ptr_global)(1) = -(x - x_ref)*sin(theta_ref) + (y - y_ref)*cos(theta_ref);
    (*ptr_global)(2) = cos(theta - theta_ref)/(-kappa_ref*(-(x - x_ref)*sin(theta_ref) + (y - y_ref)*cos(theta_ref)) + 1);
  }));
}

std::shared_ptr<Dynamics> TrackerModel::createDynamics(OcpConfig::IntegratorType type) const {
  return std::make_shared<TrackerModelDynamics>(type);
}

std::shared_ptr<CostFunction> TrackerModel::createCostFunction() const {
  return std::make_shared<TrackerModelCost<StageType::NORMINAL>>();
}

std::shared_ptr<CostFunction> TrackerModel::createTerminalCostFunction() const {
  return std::make_shared<TrackerModelCost<StageType::TERMINAL>>();
}

std::shared_ptr<Constraints> TrackerModel::createConstraint() const {
  return std::make_shared<TrackerModelConstraints<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> TrackerModel::createInitialConstraint() const {
  return std::make_shared<TrackerModelConstraints<StageType::INITIAL>>();
}

std::shared_ptr<Constraints> TrackerModel::createTerminalConstraint() const {
  return std::make_shared<TrackerModelConstraints<StageType::TERMINAL>>();
}

std::shared_ptr<Constraints> TrackerModel::createStateOnlyEqualities() const {
  return std::make_shared<TrackerModelStateOnlyEqualities<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> TrackerModel::createInitialStateOnlyEqualities() const {
  return std::make_shared<TrackerModelStateOnlyEqualities<StageType::INITIAL>>();
}

std::shared_ptr<Constraints> TrackerModel::createTerminalEqualities() const {
  return std::make_shared<TrackerModelStateOnlyEqualities<StageType::TERMINAL>>();
}

std::shared_ptr<IPMHelper> TrackerModel::createBaseIpmHelper() const {
  IPMHelper::Creater creater{
    {{StageType::NORMINAL, OcpConfig::FORWARD_EULER},[]() { return std::make_shared<TrackerModelIpmEvaluator<StageType::NORMINAL, OcpConfig::FORWARD_EULER>>(); }},
    {{StageType::NORMINAL, OcpConfig::ERK4},[]() { return std::make_shared<TrackerModelIpmEvaluator<StageType::NORMINAL, OcpConfig::ERK4>>(); }},
    {{StageType::INITIAL, OcpConfig::FORWARD_EULER},[]() { return std::make_shared<TrackerModelIpmEvaluator<StageType::INITIAL, OcpConfig::FORWARD_EULER>>(); }},
    {{StageType::INITIAL, OcpConfig::ERK4},[]() { return std::make_shared<TrackerModelIpmEvaluator<StageType::INITIAL, OcpConfig::ERK4>>(); }},
    {{StageType::TERMINAL, OcpConfig::FORWARD_EULER},[]() { return std::make_shared<TrackerModelIpmEvaluatorTerminal>(); }},
    {{StageType::TERMINAL, OcpConfig::ERK4},[]() { return std::make_shared<TrackerModelIpmEvaluatorTerminal>(); }}};
  return std::make_shared<IPMHelper>(std::move(creater));
}

REGIST_MODEL(TrackerModel)
} // namespace gpal::pnc::planning

File: src/reference_code/world_model/model/tracker_model.py
import sys

import numpy as np
import sympy as sym
from sympy import *

sys.path.append("../../pnc_solver/scripts/")

from core.ocp_model import *

file_path = os.path.abspath(__file__)
dir_path = os.path.dirname(file_path)
solver_path = os.path.dirname(dir_path)
gen_dir = solver_path + "/model/tracker_model/"
if len(sys.argv) > 1:

    gen_dir = sys.argv[1]

ocp_model = OptimalControlProblem("TrackerModel")

# define state symbols
# s 为投影到参考线上的匹配点的local s信息
s, x, y, theta, steer = ocp_model.add("state", "s, x, y, theta, steer")

# define control symbols
dsteer= ocp_model.add("ctrl", "dsteer")

# define slack symbols
l_slack = ocp_model.add("slack", "l_slack")

# reference line state params
x_ref, y_ref, theta_ref, kappa_ref = ocp_model.add(
    "param", "x_ref, y_ref, theta_ref, kappa_ref"
)

#  coarse state params s_coarse 为决策轨迹在参考线上的投影点的local s信息
v_ref = ocp_model.add("param", "v_ref")

#  terminal state params
l_terminal = ocp_model.add(
    "param", "l_terminal"
)


wheelbase, front_overhang, rear_overhang, length, width = ocp_model.add(
    "param",
    wheelbase=3.0,
    front_overhang=1.0,
    rear_overhang=1.0,
    length=5.0,
    width=2.0,
)


l_ref_weight, theta_ref_weight, steer_weight, dsteer_weight = (
    ocp_model.add(
        "param",
        l_ref_weight=1.0,
        theta_ref_weight=1.0,
        steer_weight=1.0,
        dsteer_weight=1.0,
    )
)

l_slack_weight = ocp_model.add("param", l_slack_weight=1.0)

terminal_l_weight, terminal_theta_weight = ocp_model.add(
    "param", terminal_l_weight=1.0, terminal_theta_weight=1.0
)

# constraints params
LHardLowerBound, LHardUpperBound = ocp_model.add(
    "param", LHardLowerBound=-20.0, LHardUpperBound=20.0
)

SteerLowerBound, SteerUpperBound = ocp_model.add(
    "param", SteerLowerBound=-0.436, SteerUpperBound=0.436
)  # 默认最大轮转角度为25度

DSteerLowerBound, DSteerUpperBound = ocp_model.add(
    "param", DSteerLowerBound=-0.436, DSteerUpperBound=0.436
)  # 默认最大轮转角速率为25度每秒



l = -sin(theta_ref) * (x - x_ref) + cos(theta_ref) * (y - y_ref)
ds = cos(theta - theta_ref) / (1 - kappa_ref * l)
kappa = tan(steer) / wheelbase

# define dynamics
ocp_model.subject(dot(s) == v_ref * ds)
ocp_model.subject(dot(x) == v_ref * cos(theta))
ocp_model.subject(dot(y) == v_ref * sin(theta))
ocp_model.subject(dot(theta) == v_ref * kappa)
ocp_model.subject(dot(steer) == dsteer)


(
    l_scale,
    theta_scale,
    steer_scale,
    dsteer_scale,
) = ocp_model.add(
    "param",
    l_scale=1.0,
    theta_scale=1.0,
    steer_scale=1.0,
    dsteer_scale=1.0,
)

# define cost
cost_x_lateral = (
    0.5 * l_ref_weight * (l / l_scale) ** 2
    + 0.5 * steer_weight * (steer / steer_scale) ** 2
    + 0.5 * theta_ref_weight * ((theta - theta_ref) / theta_scale) ** 2
)

cost_u = (
    0.5 * dsteer_weight * (dsteer / dsteer_scale) ** 2
)

cost_s = 0.5 * l_slack_weight * (l_slack / l_scale) ** 2 

ocp_model.add(
    "cost", cost_x_lateral, cost_u, cost_s
)

# define terminal cost
terminal_cost_x = (
    0.5 * terminal_l_weight * (l / l_scale) ** 2
    + 0.5 * terminal_theta_weight * ((theta - theta_ref) / theta_scale) ** 2
)
ocp_model.add("cost", terminal_cost_x, type=StageType.TERMINAL)

# define constraints

ocp_model.subject(LHardLowerBound <= l + l_slack, l - l_slack<= LHardUpperBound)

ocp_model.subject(SteerLowerBound <= steer, steer <= SteerUpperBound)

ocp_model.subject(DSteerLowerBound <= dsteer, dsteer <= DSteerUpperBound)

ocp_model.subject(cos(theta - theta_ref) >= 0.0)

ocp_model.generate(gen_dir)
